keras_cv/models/segmentation/deeplab_v3/deeplab_v3.py

```diff
@@ -0,0 +1,255 @@
+            specified.
+        spatial_pyramid_pooling: (Optional) a `keras.layers.Layer`. Also known
+            as Atrous Spatial Pyramid Pooling (ASPP). Performs spatial pooling
+            on different spatial levels in the pyramid, with dilation. If
+            provided, the feature map from the backbone is passed to it inside
+            the DeepLabV3 Encoder, otherwise
+            `keras_cv.layers.spatial_pyramid.SpatialPyramidPooling` is used.
+        segmentation_head: (Optional) a `keras.layers.Layer`. If provided, the
+            outputs of the DeepLabV3 encoder is passed to this layer and it
+            should predict the segmentation mask based on feature from backbone
+            and feature from decoder, otherwise a similar architecture is used.
+
+    Examples:
+    ```python
+    import tensorflow as tf
+    import keras_cv
+
+    images = tf.ones(shape=(1, 96, 96, 3))
+    labels = tf.zeros(shape=(1, 96, 96, 1))
+    backbone = keras_cv.models.ResNet50V2Backbone(input_shape=[96, 96, 3])
+    model = keras_cv.models.segmentation.DeepLabV3(
+        num_classes=1, backbone=backbone,
+    )
+
+    # Evaluate model
+    model(images)
+
+    # Train model
+    model.compile(
+        optimizer="adam",
+        loss=keras.losses.BinaryCrossentropy(from_logits=False),
+        metrics=["accuracy"],
+    )
+    model.fit(images, labels, epochs=3)
+    ```
+    """
+
+    def __init__(
+        self,
+        num_classes,
+        backbone,
+        use_low_level_features=False,
+        projection_filters=48,
+        spatial_pyramid_pooling=None,
+        segmentation_head=None,
+        dropout=0.0,
+        **kwargs,
+    ):
+        if not isinstance(backbone, keras.layers.Layer) or not isinstance(
+            backbone, keras.Model
+        ):
+            raise ValueError(
+                "Argument `backbone` must be a `keras.layers.Layer` instance "
+                f" or `keras.Model`. Received instead "
+                f"backbone={backbone} (of type {type(backbone)})."
+            )
+
+        inputs = backbone.input
+        input_shape = backbone.input_shape[1:]
+        input_tensor = backbone.input_tensor
+
+        if input_shape[0] is None and input_shape[1] is None:
+            input_shape = backbone.input_shape[1:]
+            inputs = keras.Input(tensor=input_tensor, shape=input_shape)
+
+        if input_shape[0] is None and input_shape[1] is None:
+            raise ValueError(
+                "Input shapes for both the backbone and DeepLabV3 cannot be "
+                "`None`. Received: input_shape={input_shape} and "
+                "backbone.input_shape={backbone.input_shape[1:]}"
+            )
+
+        height = input_shape[0]
+        width = input_shape[1]
+
+        feature_extractor = keras.Model(
+            inputs=backbone.input,
+            outputs=backbone.get_layer(
+                list(backbone.pyramid_level_inputs.values())[-1]
```

Comment:
Do we always want to get the lowest-resolution values here, or do we always want to take P4?

Reply:
If we want to take the lowest resolution values, let's say

```
final_backbone_pyramid_output = backbone.pyramid_level_inputs.keys()[-1]
```

and use this to build the feature extractor.

That way, when we build the upsampling later, we can say

```
upsampling_size = int(final_backbone_pyramid_output[-1]) ** 2 / 4)
encoder_outputs = keras.layers.UpSampling2D(
            size=upsampling_size,
            interpolation="bilinear",
        )(spp_outputs)
```

Reply:
If we always want P4, then it's just `upsampling_size=4`

Reply:
(benefit of both of these is that we no longer need to keep track of the input size (height+width) here, which makes the code a lot cleaner